From a84c9e125556fa921beb69c356f657df824aa0c8 Mon Sep 17 00:00:00 2001
From: chenyongcheng_H <chenyongcheng4@h-partners.com>
Date: Mon, 11 May 2026 18:53:28 +0800
Subject: [PATCH] urma: Wr_buf supports selection from either jfc or jetty
 during urma_poll_lfc

(cherry picked from commit 628282115e4e761325c503b5bf05552e77d11ee3)
---
 src/urma/lib/urma/bond/bondp_datapath.c | 69 ++++++++++++++++++++++---
 1 file changed, 63 insertions(+), 6 deletions(-)

diff --git a/src/urma/lib/urma/bond/bondp_datapath.c b/src/urma/lib/urma/bond/bondp_datapath.c
index 898cbfcf..7ac1d65c 100644
--- a/src/urma/lib/urma/bond/bondp_datapath.c
+++ b/src/urma/lib/urma/bond/bondp_datapath.c
@@ -39,6 +39,44 @@ static urma_jetty_id_t *get_comp_urma_jetty_id(bondp_comp_t *bdp_comp)
     }
 }
 
+static wr_buf_t *get_recv_wr_buf(bondp_comp_t *bdp_comp)
+{
+    if (bdp_comp == NULL) {
+        return NULL;
+    }
+    if (bdp_comp->comp_type == BONDP_COMP_JETTY) {
+        urma_jfr_t *jfr = bdp_comp->v_jetty.jetty_cfg.shared.jfr;
+        if (jfr == NULL) {
+            URMA_LOG_ERR("JETTY shared jfr is NULL\n");
+            return NULL;
+        }
+        bondp_comp_t *bdp_jfr = CONTAINER_OF_FIELD(jfr, bondp_comp_t, v_jfr);
+        return &bdp_jfr->recv_wr_buf;
+    } else if (bdp_comp->comp_type == BONDP_COMP_JFR) {
+        return &bdp_comp->recv_wr_buf;
+    }
+    return NULL;
+}
+
+static pthread_spinlock_t *get_recv_wr_lock(bondp_comp_t *bdp_comp)
+{
+    if (bdp_comp == NULL) {
+        return NULL;
+    }
+    if (bdp_comp->comp_type == BONDP_COMP_JETTY) {
+        urma_jfr_t *jfr = bdp_comp->v_jetty.jetty_cfg.shared.jfr;
+        if (jfr == NULL) {
+            URMA_LOG_ERR("JETTY shared jfr is NULL\n");
+            return NULL;
+        }
+        bondp_comp_t *bdp_jfr = CONTAINER_OF_FIELD(jfr, bondp_comp_t, v_jfr);
+        return &bdp_jfr->recv_wr_lock;
+    } else if (bdp_comp->comp_type == BONDP_COMP_JFR) {
+        return &bdp_comp->recv_wr_lock;
+    }
+    return NULL;
+}
+
 static urma_status_t comp_post_send(bondp_comp_t *comp, int send_idx, urma_jfs_wr_t *send_wr, urma_jfs_wr_t **bad_wr)
 {
     urma_status_t ret;
@@ -472,9 +510,16 @@ static urma_status_t bondp_post_recv_wr_and_store(bondp_comp_t *bdp_comp, urma_j
         return URMA_FAIL;
     }
 
-    (void)pthread_spin_lock(&bdp_comp->recv_wr_lock);
-    jfr_wr_entry_t *wr_entry = jfr_wr_buf_alloc(&bdp_comp->recv_wr_buf);
-    (void)pthread_spin_unlock(&bdp_comp->recv_wr_lock);
+    wr_buf_t *recv_wr_buf = get_recv_wr_buf(bdp_comp);
+    pthread_spinlock_t *recv_wr_lock = get_recv_wr_lock(bdp_comp);
+    if (recv_wr_buf == NULL || recv_wr_lock == NULL) {
+        URMA_LOG_ERR("Failed to get recv_wr_buf or recv_wr_lock, comp_type:%d\n", bdp_comp->comp_type);
+        return URMA_EINVAL;
+    }
+
+    (void)pthread_spin_lock(recv_wr_lock);
+    jfr_wr_entry_t *wr_entry = jfr_wr_buf_alloc(recv_wr_buf);
+    (void)pthread_spin_unlock(recv_wr_lock);
     if (wr_entry == NULL) {
         URMA_LOG_ERR("Failed to allocate jfr wr entry\n");
         return URMA_EAGAIN;
@@ -507,7 +552,9 @@ static urma_status_t bondp_post_recv_wr_and_store(bondp_comp_t *bdp_comp, urma_j
 
 FREE_PWR:
     free_jfr_wr(pwr);
+    (void)pthread_spin_lock(recv_wr_lock);
     jfr_wr_buf_release(wr_entry);
+    (void)pthread_spin_unlock(recv_wr_lock);
     return ret;
 }
 
@@ -601,8 +648,6 @@ urma_status_t bondp_post_jfr_wr(urma_jfr_t *jfr, urma_jfr_wr_t *wr, urma_jfr_wr_
             }
             cur = cur->next;
         }
-
-        return URMA_SUCCESS;
     }
     PERF_PROFILING_END(BOND_POST_JFR_RECV);
 
@@ -835,7 +880,15 @@ static cr_convert_ret_t handle_recv_cr_with_store(bondp_context_t *bdp_ctx, int
     }
 
     const uint64_t wr_id = cr->user_ctx;
-    jfr_wr_entry_t *wr_entry = jfr_wr_buf_get(&recv_comp->recv_wr_buf, wr_id);
+    wr_buf_t *recv_wr_buf = get_recv_wr_buf(recv_comp);
+    pthread_spinlock_t *recv_wr_lock = get_recv_wr_lock(recv_comp);
+    if (recv_wr_buf == NULL || recv_wr_lock == NULL) {
+        URMA_LOG_ERR("Failed to get recv_wr_buf or recv_wr_lock, comp_type:%d\n", recv_comp->comp_type);
+        put_comp(recv_comp);
+        return CONVERT_FAIL;
+    }
+
+    jfr_wr_entry_t *wr_entry = jfr_wr_buf_get(recv_wr_buf, wr_id);
     if (wr_entry == NULL) {
         // wr_entry could not be NULL
         put_comp(recv_comp);
@@ -858,7 +911,9 @@ static cr_convert_ret_t handle_recv_cr_with_store(bondp_context_t *bdp_ctx, int
     ret = bondp_conn_table_get_or_create(&recv_comp->v_conn_table, &target_jetty_id, &v_conn);
     if (ret != 0) {
         free_jfr_wr(&wr_entry->wr);
+        (void)pthread_spin_lock(recv_wr_lock);
         jfr_wr_buf_release(wr_entry);
+        (void)pthread_spin_unlock(recv_wr_lock);
         put_comp(recv_comp);
         return CONVERT_FAIL;
     }
@@ -875,7 +930,9 @@ static cr_convert_ret_t handle_recv_cr_with_store(bondp_context_t *bdp_ctx, int
     (void)bdp_slide_wnd_add(&v_conn->recv_wnd, msn);
 
     free_jfr_wr(&wr_entry->wr);
+    (void)pthread_spin_lock(recv_wr_lock);
     jfr_wr_buf_release(wr_entry);
+    (void)pthread_spin_unlock(recv_wr_lock);
     put_comp(recv_comp);
     return CONVERT_SUCCESS;
 }
-- 
2.43.0

